knapsack/knapsack.py

```diff
@@ -1,43 +1,68 @@
-""" A naive recursive implementation of 0-1 Knapsack Problem
+""" A recursive implementation of 0-N Knapsack Problem
     https://en.wikipedia.org/wiki/Knapsack_problem
 """
 from __future__ import annotations
 
+from functools import lru_cache
 
-def knapsack(capacity: int, weights: list[int], values: list[int], counter: int) -> int:
+
+def knapsack(
+    capacity: int,
+    weights: list[int],
+    values: list[int],
+    counter: int,
+    allow_repetition=False,
+) -> int:
     """
     Returns the maximum value that can be put in a knapsack of a capacity cap,
-    whereby each weight w has a specific value val.
+    whereby each weight w has a specific value val
+    with option to allow repetitive selection of items
 
     >>> cap = 50
     >>> val = [60, 100, 120]
     >>> w = [10, 20, 30]
     >>> c = len(val)
-    >>> knapsack(cap, w, val, c)
+    >>> knapsack(cap, w, val, c, False)
     220
 
-    The result is 220 cause the values of 100 and 120 got the weight of 50
+    Given the repetition is NOT allowed,
+    the result is 220 cause the values of 100 and 120 got the weight of 50
+    which is the limit of the capacity.
+    >>> knapsack(cap, w, val, c, True)
+    300
+
+    Given the repetition is allowed,
+    the result is 300 cause the values of 60*5 (pick 5 times)
```

Comment:
The algorithm has been updated to allow repetition. An example/testcase where repetition is used would be appropriate.